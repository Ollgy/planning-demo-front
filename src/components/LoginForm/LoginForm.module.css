.formContainer {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  width: 100%;
  height: 100%;
  justify-content: center;
  align-items: center;
}

.formWrap {
  position: relative;
  width: 365px;
  padding: 45px 35px 25px 35px;
  border: 2px solid #bfbfbf;
  border-radius: 15px;
  box-shadow: 5px 10px 10px 2px #e6e6e6, -5px 0px 10px 2px #e6e6e6;
  background-color: #fff;
  overflow: hidden;
}

.formWrap:after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 35px;
  background: linear-gradient(#e6e6e6, #fff);
}

.form {
  width: 100%;
}

.form * {
  font-size: 16px;
}

.formItem {
  display: flex;
  flex-direction: row-reverse;
  min-height: 35px;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 1vh;
}

.formItem span {
  flex: 1;
}

.formItemColumn {
  flex-direction: column-reverse;
  justify-content: flex-start;
  align-items: flex-start;
}

.formItemColumn > label {
  margin-bottom: 10px;
}

.formItemWide {
  margin-bottom: 20px;
}

.formItem:last-of-type {
  margin-bottom: 0;
}

.formLabel {
  flex: 1;
  color: #757575;
  margin-right: 25px;
}

.formValue {
  flex: 1;
}

.msgBlock {
  height: 20px;
}

.buttonContainer {
  width: 100%;
  padding-top: 25px;
  text-align: center;
}
