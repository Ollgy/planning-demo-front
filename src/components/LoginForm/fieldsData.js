const fieldsData = [
  {
    name: 'login',
    type: 'text',
    label: 'Логин'
  },
  {
    name: 'password',
    type: 'password',
    label: 'Пароль'
  }
];

export default fieldsData;
