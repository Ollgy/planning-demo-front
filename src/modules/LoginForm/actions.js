import { createAction } from 'redux-actions';

export const validateForm = createAction('VALIDATE_FORM_LOGIN');
export const resetForm = createAction('RESET_FORM_LOGIN');
export const inputForm = createAction('INPUT_FORM_LOGIN');
export const toggleChb = createAction('TOGGLE_CHB_LOGIN');
export const clearError = createAction('CLEAR_ERROR_LOGIN');
export const clearFields = createAction('CLEAR_FIELDS_LOGIN');
