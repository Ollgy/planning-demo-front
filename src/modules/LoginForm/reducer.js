import {
  validateForm,
  resetForm,
  inputForm,
  toggleChb,
  clearFields,
  clearError
} from './actions';
import validate from './utils/validate';
import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';

export const initialFields = {
  login: '',
  password: '',
  remembered: false
};

const initialValidation = {
  status: false,
  msg: '',
  error: '',
  errorInputs: []
};

const fields = handleActions(
  {
    [resetForm]: () => initialFields,

    [clearFields]: (_state, action) =>
      Object.keys(initialFields).reduce(
        (obj, prop) => ({
          ...obj,
          [prop]: action.payload[prop] ? initialFields[prop] : _state[prop]
        }),
        {}
      ),

    [inputForm]: (_state, action) => ({
      ..._state,
      [action.payload.input]: action.payload.value
    }),

    [toggleChb]: _state => ({ ..._state, remembered: !_state.remembered })
  },
  initialFields
);

const validation = handleActions(
  {
    [validateForm]: (_state, action) => validate(action.payload),

    [clearError]: (_state, action) => {
      const nextErrInputs = _state.errorInputs.filter(
        input => input !== action.payload.input
      );

      return {
        ..._state,
        errorInputs: nextErrInputs,
        msg: nextErrInputs.length ? _state.msg : ''
      };
    },

    [resetForm]: () => initialValidation
  },
  initialValidation
);

export default combineReducers({
  fields,
  validation
});
