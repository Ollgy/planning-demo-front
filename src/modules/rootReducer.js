import { combineReducers } from 'redux';
import authFromToken from './Auth';
import login from './Login';
import loginForm from './LoginForm';
import profile from './Profile';
import personalForm from './PersonalForm';
import changePswrdForm from './ChangePswrdForm';
import newTaskForm from './NewTaskForm';
import notes from './Notes';
import dashboard from './Dashboard';
import staff from './Staff';
import cleanBase from './CleanBase';

export default combineReducers({
  authFromToken,
  login,
  loginForm,
  profile,
  personalForm,
  changePswrdForm,
  newTaskForm,
  notes,
  dashboard,
  staff,
  cleanBase
});
