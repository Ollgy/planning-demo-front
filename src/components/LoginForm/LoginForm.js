import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import Input from '../Input';
import Button from '../Button';
import FormMsg from '../FormMsg';
import Checkbox from '../Checkbox';
import fieldsData from './fieldsData';
import { login } from '../../modules/Login';
import {
  inputForm,
  resetForm,
  validateForm,
  toggleChb,
  clearError,
  clearFields
} from '../../modules/LoginForm';
import { F } from '../../const';
import styles from './LoginForm.module.css';

class LoginForm extends PureComponent {
  constructor(props) {
    super(props);
    this.activeInput = React.createRef();
  }

  componentDidMount() {
    this.props.resetForm();
    this.activeInput.current.focus();
  }

  onInputChange = e => {
    const { inputForm, clearError } = this.props;

    inputForm({ input: e.target.name, value: e.target.value });
    clearError({ input: e.target.name });
  };

  onSubmit = async e => {
    e.preventDefault();
    const { fields, validateForm, clearFields, login, user } = this.props;

    await validateForm({ fields, progress: 'prepare' });

    const { validation } = this.props;

    if (validation.status) {
      const { login: email, password, remembered } = fields;
      await login({ email, password, remembered });

      if (!user) {
        validateForm({ fields, progress: 'access_error' });
        clearFields({ login: email, password });
      } else {
        validateForm({ fields, progress: 'access_success' });
      }
    }

    if (this.activeInput.current) {
      this.activeInput.current.focus();
    }
  };

  render() {
    const { fields, resetForm, toggleChb } = this.props;
    const { error, msg, errorInputs } = this.props.validation;

    return (
      <div className={styles.formContainer}>
        <div className={styles.formWrap}>
          <form
            action="/"
            method="post"
            className={styles.form}
            onSubmit={this.onSubmit}
            onReset={resetForm}
          >
            {fieldsData.map((field, i) => (
              <div
                className={`${styles.formItem} ${styles.formItemWide}`}
                key={field.name}
              >
                <Input
                  type={field.type}
                  name={field.name}
                  ref={
                    errorInputs.length
                      ? field.name === errorInputs[0]
                        ? this.activeInput
                        : null
                      : i === 0
                      ? this.activeInput
                      : null
                  }
                  value={fields[field.name]}
                  placeholder=". . ."
                  errorClass={errorInputs.includes(field.name) ? error : null}
                  onChange={this.onInputChange}
                />
                <label className={styles.formLabel} htmlFor={field.name}>
                  <span>{field.label}</span>
                </label>
              </div>
            ))}
            <div className={`${styles.formItem}`} key={F.remembered}>
              <Checkbox
                type="checkbox"
                id={F.remembered}
                value={F.remembered}
                name={F.remembered}
                checked={fields.remembered}
                onChange={toggleChb}
              />
              <label className={styles.formLabel} htmlFor={F.remembered}>
                <span>Запомнить меня</span>
              </label>
            </div>
            <div className={styles.msgBlock}>
              <FormMsg children={msg} />
            </div>
            <div className={styles.buttonContainer}>
              <Button type="submit" name="submit" children="Войти" />
            </div>
          </form>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  fields: state.loginForm.fields,
  validation: state.loginForm.validation,
  user: state.login.user,
  isLoading: state.login.isLoading,
  error: state.login.error
});

const mapDispatchToProps = dispatch => ({
  login: arg => dispatch(login(arg)),
  validateForm: arg => dispatch(validateForm(arg)),
  inputForm: arg => dispatch(inputForm(arg)),
  resetForm: () => dispatch(resetForm()),
  toggleChb: () => dispatch(toggleChb()),
  clearError: arg => dispatch(clearError(arg)),
  clearFields: arg => dispatch(clearFields(arg))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginForm);
