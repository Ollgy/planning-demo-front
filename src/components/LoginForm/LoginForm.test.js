import React from 'react';
import LoginForm from './LoginForm';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { createStore, compose, applyMiddleware } from 'redux';
import rootReducer from '../../modules/rootReducer';
import thunk from 'redux-thunk';

const store = createStore(rootReducer, {}, compose(applyMiddleware(thunk)));

describe('<LoginForm/> UI component', () => {
  const shallowExpect = compose(
    expect,
    toJSON,
    shallow
  );

  it('Renders correctly', () => {
    shallowExpect(<LoginForm store={store} />).toMatchSnapshot();
  });

  const wrapper = shallow(<LoginForm store={store} />);
  const inner = wrapper.children().first();

  it('contains fields in props', () => {
    expect(inner.props().fields).toBeTruthy();
  });

  it('contains validation in props', () => {
    expect(inner.props().validation).toBeTruthy();
  });
});
