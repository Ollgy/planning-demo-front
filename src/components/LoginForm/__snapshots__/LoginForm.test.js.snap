// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm/> UI component Renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <LoginForm
    clearError={[Function]}
    clearFields={[Function]}
    error={null}
    fields={
      Object {
        "login": "",
        "password": "",
        "remembered": false,
      }
    }
    inputForm={[Function]}
    isLoading={false}
    login={[Function]}
    resetForm={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    toggleChb={[Function]}
    user={null}
    validateForm={[Function]}
    validation={
      Object {
        "error": "",
        "errorInputs": Array [],
        "msg": "",
        "status": false,
      }
    }
  />
</ContextProvider>
`;
